samza-core/src/main/java/org/apache/samza/task/StreamOperatorTask.java

```diff
@@ -100,8 +103,11 @@ public final void process(IncomingMessageEnvelope ime, MessageCollector collecto
   }
 
   @Override
-  public final void window(MessageCollector collector, TaskCoordinator coordinator) throws Exception {
-    // TODO: invoke timer based triggers
+  public final void window(MessageCollector collector, TaskCoordinator coordinator)  {
+    context.getSystemStreamPartitions().forEach(ssp -> {
+        OperatorImpl impl = this.operatorGraph.get(ssp.getSystemStream());
+        impl.onTick(collector, coordinator);
```

Comment:
I wonder that it would help if we allow passing in a clock function in onTick that always returns the nanoSecond when window() is invoked, to avoid the potential perf issue in calling System.nanoTime() multiple times in each operator.onTimer().